### ISSUE TITLE
Folder thumbnails including folder.jpg support

### ISSUE THREAD

**Issue by ReaperX** [mate-desktop/caja#751]
Many people make their media collections both better looking and easier to manage by placing a folder.jpg cover picture into each subfolder. A thumbnail of this picture is shown inside the folder icon. It would be wonderful if Caja supported this functionality in icon view.

In my opinion, the most visually appealing way of implementing this would be Windows XP style, see

http://i37.tinypic.com/2mf0m6f.jpg

where a two-dimensional thumbnail of the folder.jpg is put inside a thin standard folder frame. 

Absent a folder.jpg, (but only then) Caja would show several (up to 4) smaller thumbnails of media files contained in the folder.

**Comment by ReaperX** [comment]
Just wanted to add that a solution already exists for Nautilus.

http://projects.flogisoft.com/cover-thumbnailer/

Can this be adjusted to work with Caja?

**Comment by samuel0l0y** [comment]
The solution above is kind of a hack, because it stores the images in the hard disk, and so will have to be run again for every new folder.

**Comment by davidedelvento** [comment]
I have been doing _something_ (more on that in a bit) which made folders behave like that for more than 5 years. Yesterday I reinstalled the OS on an old computer and I could not remember what it was. Searched for an hour for a solution and found many people complaining on various forums that this functionality does not exist. Found a few suggestions (including this one) about cover-thumbnailer which did not work on that version of MATE that my new OS provides.

If I did not have remembered **for sure** that it **absolutely did** work in the past, I would have given up. Then, the first clue: if you right-click on a folder and select Properties from that menu, you get a window with a tabbed interface, preselected in the Basic tab. You can (left-)click on the folder icon and select an icon of your choice, and that works just fine. Yet, you don't want to do that for thousands of folders, but maybe one can script that? I changed a few icons, and I `find`'ed all files (including in hidden directories) in my home directory `xargs`'ing `grep` the filenames of the icons. No luck, this information is stored in some obscure way (perhaps in some compressed format of some database).

Yet I did remember that **it did work**, so I persisted. And found the way, only because my former self knew that my current self would have banged the head that hard on the keyboard, so that kind soul wrote a note somewhere for me. I just wish he made that note a bit more discoverable, so I would have spent less time. But I am thanking him nevertheless. And I am going to tell you the solution in a moment. You may wonder why all this preamble, and the reason is that this feature is implemented in such an obscure way, that nobody on the internet knows about it. Maybe there needs to be a tiny UI change to make it more discoverable, and **most definitely** it needs to be documented.

Anyway, the programmatic way to set the custom folder icon is with the `gio` commandline too to set the `custom-icon` metadata.

```bash
for file in /path/of/your/interest/*/*/folder.jpg; do
    gio set $(dirname $file) metadata::custom-icon $(basename $file);
done
```

That ran in a blink on the hundreds of directories I had. 

I am not going to do other than mention some "obvious" facts here:

* This will work if you have a flat hierarchy like I do. Otherwise you better do `find /path/of/your/interest/ -name folder.jpg | xargs -I {} gio set $(dirname {}) metadata::custom-icon $(basename {})` (untested, but should work -- see [here](https://unix.stackexchange.com/questions/267437/how-to-repeat-variables-twice-in-xargs) for details)
* Neither the `for` nor the `find` approach will work with trees with spaces. I think files and directories with spaces are evil and should never be used, and I never do. If you insist, learn how to properly escape them, but don't ask me. You might start [here](https://stackoverflow.com/questions/16758525/make-xargs-handle-filenames-that-contain-spaces)
* You can use different name than `folder.jpg` and even a more sophisticated rule such that some folders get some sort of icon and some others get a different one. This is really cool

So in conclusion, this functionality is already implemented in caja and has been for a long time. It just need to be a UI to make it easy for people to use it, and documentation for users (but most importantly developers and administrators) to be able to script it. 

A simple UI option could be a button under the current folder icon in the Properties/Basic. Such a button could do something like "use `folder.jpg` as icon for this directory and all subdirectories" and it could simply run a variant of the script above.

HTH
